src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/Streaming/DataStreamWriterTests.cs

```diff
@@ -59,5 +63,93 @@ public void TestSignaturesV2_3_X()
 
             Assert.IsType<DataStreamWriter>(dsw.Trigger(Trigger.Once()));
         }
+
+        [SkipIfSparkVersionIsLessThan(Versions.V2_4_0)]
+        public void TestForeach()
+        {
+            // Temporary folder to put our test stream input.
+            using var srcTempDirectory = new TemporaryDirectory();
+            string streamInputPath = Path.Combine(srcTempDirectory.Path, "streamInput");
+
+            // [1, 2, ..., 99]
+            _spark.Range(1, 100).Write().Json(streamInputPath);
+
+            // Temporary folder the TestForeachWriter will write to.
+            using var dstTempDirectory = new TemporaryDirectory();
+            int partitions = 3;
+
+            // Read streamInputPath, repartitions data into 3, then
+            // calls TestForeachWriter on the data.
+            DataStreamWriter dsw = _spark
+                .ReadStream()
+                .Schema("id INT")
+                .Json(streamInputPath)
+                .Repartition(partitions)
+                .WriteStream()
+                .Foreach(new TestForeachWriter(dstTempDirectory.Path));
+
+            // Trigger the stream batch once.
+            dsw.Trigger(Trigger.Once()).Start().AwaitTermination();
+
+            // Verify that TestForeachWriter wrote a unique .csv for each
+            // partition.
+            Assert.Equal(
+                partitions,
+                Directory.GetFiles(dstTempDirectory.Path, "*.csv").Length);
+
+            // Read in the *.csv file(s) generated by the TestForeachWriter.
+            // If there are multiple input files, sorting by "id" will make
+            // validation simpler.
+            DataFrame foreachWriterOutputDF = _spark
+                .Read()
+                .Schema("id INT")
+                .Csv(dstTempDirectory.Path)
+                .Sort("id");
+
+            // [101, 102, ..., 199]
+            IEnumerable<int> expected = Enumerable.Range(101, 99);
+
+            Assert.Equal(
+                expected.Select(i => new object[] { i }),
+                foreachWriterOutputDF.Collect().Select(r => r.Values));
+        }
+
+        [Serializable]
+        private class TestForeachWriter : IForeachWriter
+        {
+            [ThreadStatic]
+            private static StreamWriter s_streamWriter;
```

Comment:
I think this  may not be as important for Windows or the unit tests, but if this was running on Linux and if the worker was running multiple `TestForeachWriter` tasks in parallel, then they may end up using the same `StreamWriter` and write to the same file instead of different files.

Reply:
can you capture your intention in the comment?

Reply:
Added to comments.